### PR TITLE
ENH: Improve usability of Directory Button by using native dialog

### PR DESCRIPTION
**Pull request by jamesobutler**
This aims to improve directory selection by using the native dialog instead of Qt's own widget-based dialog design.

See 3D Slicer forum discussion of how users have been doing other methods to avoid using widgets that use Qt's widget-based dialog.
https://discourse.slicer.org/t/using-native-file-dialog-for-selecting-files-and-folders/17926

| Qt widget-based dialog | Native dialog (on Windows)|
|-------------------------|--------------------------------|
|image|image|

### CODE REVIEW

**Comment by lassoan** [comment]
Thank you for working on this. Native file dialogs are now probably preferred by most users, so I think it is a good idea to use them by default.

Is there a reason why you switched from ctkFileDialog to QFileDialog instead of just using the `DontUseNativeDialog` flag to choose between native and CTK style? You could even change the default to use native dialog, but keep the door open for using the CTK style dialog to reduce the potential undesired impact of the change.

**Comment by jamesobutler** [comment]
I switched from ctkFileDialog to QFileDialog because the customizations of ctkFileDialog appeared to be mostly related to modifying the functionality of a Qt widget-based dialog. So switching to the native file dialog is essentially going back to using a QFileDialog without those additional customizations.

We can't just remove the following `DontUseNativeDialog` specification as there is other code in the class that requires the use of a Qt widget-based dialog with calls such as `setBottomWidget`.
https://github.com/commontk/CTK/blob/01ab373d01e6e8718af42f856278b61bb18fbcfe/Libs/Widgets/ctkFileDialog.cpp#L130-L135

**Comment by lassoan** [comment]
The extra features don't have to be deleted, but they would only be used if dontusenativedialog flag is enabled.

File: Libs/Widgets/ctkFileDialog.cpp [hunk 1]
```diff
@@ -78,6 +78,10 @@ void ctkFileDialogPrivate::init()
 QPushButton* ctkFileDialogPrivate::acceptButton()const
 {
   Q_Q(const ctkFileDialog);
+  if (q->Options)
```
[the remark this answers is not in the document]
Comment (lassoan) [reply Libs/Widgets/ctkFileDialog.cpp:81]: `if (!(q->Options & DontUseNativeDialog))`

File: Libs/Widgets/ctkFileDialog.cpp [hunk 2]
```diff
@@ -145,6 +142,10 @@ ctkFileDialog::~ctkFileDialog()
 //------------------------------------------------------------------------------
 void ctkFileDialog::setBottomWidget(QWidget* widget, const QString& label)
 {
+  if (this->Options)
```
Comment (lassoan) [inline Libs/Widgets/ctkFileDialog.cpp:145]:
```suggestion
if (!(q->Options & DontUseNativeDialog))
```

File: Libs/Widgets/ctkFileDialog.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (jamesobutler) [reply Libs/Widgets/ctkFileDialog.cpp:81]: With that change:
> Error	C7624	Type name 'QFileDialog::Options' cannot appear on the right side of a class member access expression	CTKWidgets	C:\S5R-nightly\CTK\Libs\Widgets\ctkFileDialog.cpp	81	

> Error	C2065	'DontUseNativeDialog': undeclared identifier	CTKWidgets	C:\S5R-nightly\CTK\Libs\Widgets\ctkFileDialog.cpp	81

File: Libs/Widgets/ctkFileDialog.cpp (hunk 2 above)
Comment (jamesobutler) [reply Libs/Widgets/ctkFileDialog.cpp:145]: With that change:

> Error	C2065	'q': undeclared identifier	CTKWidgets	C:\S5R-nightly\CTK\Libs\Widgets\ctkFileDialog.cpp	145

File: Libs/Widgets/ctkFileDialog.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (lassoan) [reply Libs/Widgets/ctkFileDialog.cpp:81]: Probably you need to add the scope - QFileDialog::DontUseNativeDialog

File: Libs/Widgets/ctkFileDialog.cpp (hunk 2 above)
Comment (lassoan) [reply Libs/Widgets/ctkFileDialog.cpp:145]: I haven't compiled the code. You need to use the scope wherever Options is defined in(d->, q->, this->...).

**Comment by lassoan** [comment]
Thanks for the updates. These small changes are much less likely to cause any regressions in any CTK-based projects. Just need to sort out those few syntax errors.

File: Libs/Widgets/ctkFileDialog.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (jamesobutler) [reply Libs/Widgets/ctkFileDialog.cpp]: It was the first error I mentioned that was mostly tripping me up. The documentation at https://doc.qt.io/qt-5/qfiledialog.html regarding `Options`, `options` was confusing, but I was able to get it working with `options()`. I've pushed that update.

File: Libs/Widgets/ctkFileDialog.cpp [hunk 3]
```diff
@@ -145,6 +146,10 @@ ctkFileDialog::~ctkFileDialog()
 //------------------------------------------------------------------------------
 void ctkFileDialog::setBottomWidget(QWidget* widget, const QString& label)
 {
+  if (!(this->options() & QFileDialog::DontUseNativeDialog))
+  {
+    return;  // Native dialog does not supporting modifying or getting widget elements which are done as part of init()
```
Comment (finetjul) [inline Libs/Widgets/ctkFileDialog.cpp:151]: The comment is misleading, `init()` is not being called in this function

File: Libs/Widgets/ctkFileDialog.cpp [hunk 4]
```diff
@@ -78,6 +78,10 @@ void ctkFileDialogPrivate::init()
 QPushButton* ctkFileDialogPrivate::acceptButton()const
 {
   Q_Q(const ctkFileDialog);
+  if (!(q->options() & QFileDialog::DontUseNativeDialog))
+  {
+    return NULL;  // Native dialog does not supporting modifying or getting widget elements which are done as part of init()
```
Comment (finetjul) [inline Libs/Widgets/ctkFileDialog.cpp:83]: The comment is misleading, `init()` is not being called in this function

File: Libs/Widgets/ctkFileDialog.cpp [hunk 5]
```diff
@@ -127,13 +131,10 @@ ctkFileDialog::ctkFileDialog(QWidget *parentWidget,
...
-// dialogs don't publish any internals. No problems on other OS.
-// Can be applied to Qt4 as well, if problems arise there.
-#if QT_VERSION >= QT_VERSION_CHECK(5,0,0)
-  this->setOptions(DontUseNativeDialog);
-#endif
-
+  if (!(this->options() & QFileDialog::DontUseNativeDialog))
+  {
+    return;  // Native dialog does not supporting modifying or getting widget elements which are done as part of init()
+  }
```
Comment (finetjul) [inline Libs/Widgets/ctkFileDialog.cpp:137]: I would have moved that condition in `init()` directly, in case there are (or will be added) other important operations in `init()` that are not related to to native/not native.

**Comment by jamesobutler** [comment]
I've pushed an update based on @finetjul 's suggestions

**Comment by jamesobutler** [comment]
Thoughts on latest updates @lassoan @finetjul ?

File: Libs/Widgets/ctkFileDialog.cpp [hunk 6]
```diff
@@ -67,6 +67,11 @@ void ctkFileDialogPrivate::init()
 {
   Q_Q(ctkFileDialog);
 
+  if (!(q->options() & QFileDialog::DontUseNativeDialog))
+  {
+    return;  // Native dialog does not supporting modifying or getting widget elements.
```
Comment (lassoan) [inline Libs/Widgets/ctkFileDialog.cpp:72]: I think @finetjul's suggestion was to not return but put the content that is only relevant for non-native dialog (which is currently all the content of this method) in the if branch. That way if some more features are added in the future there is smaller chance that it will be accidentally skipped.

File: Libs/Widgets/ctkFileDialog.cpp [hunk 7]
```diff
@@ -67,6 +67,11 @@ void ctkFileDialogPrivate::init()
 {
   Q_Q(ctkFileDialog);
 
+  if (!(q->options() & QFileDialog::DontUseNativeDialog))
```
Comment (lassoan) [inline Libs/Widgets/ctkFileDialog.cpp:70]: What happens if donusenativedialog flag is set after init()? If you don't want to allow changing native/non-native flag after initialization (which may be a good idea, because it is probably rarely needed and it would take some effort to implement correctly) then it would be safer to store a the value of the DontUseNativeDialog flag in a member variable and use that (instead of checking options(), which may change).

**Review by lassoan — COMMENTED** [review]
Thank you!
I had just two comments that should be easy to address.

File: Libs/Widgets/ctkFileDialog.cpp (hunk 7 above)
Comment (jamesobutler) [reply Libs/Widgets/ctkFileDialog.cpp]: I've pushed an additional commit with changes to a member variable. See if this matches what you were thinking.

**Review by lassoan — APPROVED** [review]
Thank you, this looks good to me.

File: Libs/Widgets/ctkFileDialog.cpp [hunk 8]
```diff
@@ -60,24 +61,32 @@ ctkFileDialogPrivate::ctkFileDialogPrivate(ctkFileDialog& object)
...
 {
   Q_Q(ctkFileDialog);
-
-  this->observeAcceptButton();
-
-  QObject::connect(this->listView()->selectionModel(),
-                   SIGNAL(selectionChanged(QItemSelection,QItemSelection)),
-                   q, SLOT(onSelectionChanged()));
+  this->UsingNativeDialog = !(q->options() & QFileDialog::DontUseNativeDialog);
+  if (!(this->UsingNativeDialog))
```
Comment (jcfr) [inline Libs/Widgets/ctkFileDialog.cpp:72]:
```suggestion
  if (!this->UsingNativeDialog)
```

Comment (jamesobutler) [reply Libs/Widgets/ctkFileDialog.cpp:72]: I've pushed commit with change to #977
